src/main/scala/mesosphere/marathon/core/groupmanager/impl/GroupManagerActor.scala

```diff
@@ -0,0 +1,262 @@
+package mesosphere.marathon.core.groupmanager.impl
+
+import java.net.URL
+import javax.inject.Provider
+
+import akka.actor.{ Actor, ActorLogging, Props }
+import akka.event.EventStream
+import akka.pattern.pipe
+import mesosphere.marathon._
+import mesosphere.marathon.api.v2.Validation._
+import mesosphere.marathon.event.{ GroupChangeFailed, GroupChangeSuccess }
+import mesosphere.marathon.io.PathFun
+import mesosphere.marathon.io.storage.StorageProvider
+import mesosphere.marathon.state.{ AppDefinition, Container, PortDefinition, _ }
+import mesosphere.marathon.upgrade.{ DeploymentPlan, GroupVersioningUtil, ResolveArtifacts }
+import mesosphere.util.CapConcurrentExecutions
+import org.slf4j.LoggerFactory
+
+import scala.collection.immutable.Seq
+import scala.collection.mutable
+import scala.concurrent.Future
+import scala.util.{ Failure, Success }
+
+private[groupmanager] object GroupManagerActor {
+  def props(
+    serializeUpdates: CapConcurrentExecutions,
+    scheduler: Provider[DeploymentService],
+    groupRepo: GroupRepository,
+    appRepo: AppRepository,
+    storage: StorageProvider,
+    config: MarathonConf,
+    eventBus: EventStream): Props = {
+    Props(new GroupManagerActor(
+      serializeUpdates,
+      scheduler,
+      groupRepo,
+      appRepo,
+      storage,
+      config,
+      eventBus))
+  }
+}
+
+private[impl] class GroupManagerActor(
+    serializeUpdates: CapConcurrentExecutions,
+    // a Provider has to be used to resolve a cyclic dependency between CoreModule and MarathonModule.
+    // Once MarathonSchedulerService is in CoreModule, the Provider could be removed
+    schedulerProvider: Provider[DeploymentService],
+    groupRepo: GroupRepository,
+    appRepo: AppRepository,
+    storage: StorageProvider,
+    config: MarathonConf,
+    eventBus: EventStream) extends Actor with ActorLogging with PathFun {
+  import GroupManagerDelegate._
+  import context.dispatcher
+
+  private[this] val log = LoggerFactory.getLogger(getClass.getName)
+  private var scheduler: DeploymentService = _
+
+  override def preStart(): Unit = {
+    super.preStart()
+    scheduler = schedulerProvider.get()
+  }
+
+  override def receive: Receive = {
+    Seq(
+      receiveApp,
```

Comment:
You basically translate every method into a receive function.
Multiple receive functions make sense, if you have different behaviors or complex receive functions.

- it is not easy to see which messages get handled
- you tangle messaging logic and business logic
- every receive function directly handles the logic
- it is not easy to test the domain logic

I would suggest this:

```scala
def receive: Receive = {
  case GetRootGroup => getRootGroup().pipeTp(sender())
  case GetGroupWithId(id) => getGroupWithId(id).pipeTo(sender())
}

def getRootGroup(): Future[Group] = ???

```